cmd/sync-rover-groups/yamlgroupcollector.go

```diff
@@ -0,0 +1,206 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	templatev1 "github.com/openshift/api/template/v1"
+	userv1 "github.com/openshift/api/user/v1"
+)
+
+type yamlGroupCollector struct {
+	decoder          runtime.Decoder
+	validateSubjects bool
+}
+
+const yamlSeparator = "\n---"
+
+func (c *yamlGroupCollector) collect(dir string) (sets.String, error) {
+	groups := sets.NewString()
+	abs, err := filepath.Abs(dir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine absolute path for dir %s: %w", dir, err)
+	}
+	if err := filepath.Walk(abs,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				logrus.WithField("source-file", path).WithError(err).Error("Failed to walk release controller mirror config dir")
+				return err
+			}
+
+			if skip, err := fileFilter(info, path); skip || err != nil {
+				return err
+			}
+
+			data, err := ioutil.ReadFile(path)
+			if err != nil {
+				logrus.WithField("source-file", path).WithError(err).Error("Failed to read file")
+				return err
+			}
+			for _, yamlDoc := range bytes.Split(data, []byte(yamlSeparator)) {
+				if onlyComments(yamlDoc) {
+					continue
+				}
+				obj, _, err := c.decoder.Decode(yamlDoc, nil, nil)
+				if err != nil {
+					if runtime.IsNotRegisteredError(err) && (!c.validateSubjects ||
+						!strings.HasPrefix(err.Error(), "no kind \"Group\" is registered for version \"v1\"")) {
+						continue
+					}
+					logrus.WithField("source-file", path).WithError(err).Error("Failed to decode yaml file")
+					return err
+				}
+				switch o := obj.(type) {
+				case *v1.List:
```

Comment:
Could this block be extracted to soething like `processList`?

Reply:
The current implementation is more succinct.
Maybe a new function is not that essential?

Reply:
Yes, looks good now :+1: Thanks!